leges.go

```diff
@@ -76,116 +33,142 @@ func (e *ErrExprRunFailed) Error() string {
 }
 
 type ErrExprCompileFailed struct {
-	Env     Attributes
-	Policy  Policy
-	Request Request
-	Err     error
+	Env    Attributes
+	Policy Policy
+	Err    error
 }
 
 func (e *ErrExprCompileFailed) Unwrap() error {
 	return e.Err
 }
 
 func (e *ErrExprCompileFailed) Error() string {
-	return "failed to compile expression"
+	return "failed to compileCondition expression"
 }
 
-// Leges is the struct that contains the policies and can be used to requests
-type Leges struct {
-	policies []Policy
+// statute is one policy and the compiled program of the condition
+type statute struct {
+	policy  Policy
+	program *vm.Program
 }
 
-func NewLeges(policies []Policy) (*Leges, error) {
-	visitedPolicyIDs := map[string]struct{}{}
-	for _, p := range policies {
-		if p.ID == "" {
-			return nil, ErrEmptyPolicyID
-		}
-		if _, ok := visitedPolicyIDs[p.ID]; ok {
-			return nil, fmt.Errorf("id=%q: %w", p.ID, ErrDuplicatePolicyID)
-		}
-		visitedPolicyIDs[p.ID] = struct{}{}
+// Attributes is a set of key-value attributes for objects and subjects.
+type Attributes = map[string]interface{}
+
+func New(policies []Policy, definitions Attributes) (*Leges, error) {
+	leges := &Leges{}
+
+	if err := leges.loadDefinitions(definitions); err != nil {
+		return nil, err
 	}
 
-	return &Leges{
-		policies: policies,
-	}, nil
+	if err := leges.loadPolicies(policies); err != nil {
+		return nil, err
+	}
+
+	return leges, nil
 }
 
-// Match checks a request against policies and returns whether the request matches any of the policies.
-func (lg *Leges) Match(request Request, userEnv Attributes) (bool, *Policy, error) {
-	if request.Object == nil || len(request.Object) == 0 {
-		return false, nil, ErrEmptyObjectAttrs
-	}
-	if request.Subject == nil || len(request.Subject) == 0 {
-		return false, nil, ErrEmptySubjectAttrs
+func (l *Leges) loadDefinitions(definitions Attributes) error {
+	l.definitions = Attributes{}
+
+	for k, v := range definitions {
+		l.definitions[k] = v
 	}
-	if request.Action == "" {
-		return false, nil, ErrEmptyAction
+
+	return nil
+}
+
+// loadPolicies iterates over a list of policies, runs validation on each of them and
+// keeps a compiled program of the condition for further use cases
+func (l *Leges) loadPolicies(polices []Policy) error {
+	l.statutes = make(map[string]statute, len(polices))
+
+	for _, policy := range polices {
+		if err := policy.Validate(); err != nil {
+			return err
+		}
+
+		if _, ok := l.statutes[policy.ID]; ok {
+			return fmt.Errorf("id=%q: %w", policy.ID, ErrDuplicatePolicyID)
+		}
+
+		program, err := policy.compileCondition()
+		if err != nil {
+			return &ErrExprCompileFailed{
+				Env:    l.definitions,
+				Policy: policy,
+				Err:    err,
+			}
+		}
+
+		l.statutes[policy.ID] = statute{
+			policy:  policy,
+			program: program,
+		}
 	}
 
-	env := map[string]interface{}{}
+	return nil
+}
+
+// makeClaim merges the request with the definitions to form the claim
```

Comment:
I agree, env and definitions have the same meaning and there should be only one of them. but as for the claim, in my head it's a request that's merged with userEnv. what do you think?

Reply:
I think I see what you mean, like jwt claims, right? lol

a jwt claim might be a lie/impersonation until we compute and validate the token's signature, but in Leges a Request is assumed to reflect the truth lol (it's the Leges caller's responsibility to refuse bogus claims and only allow genuine Requests)

Reply:
I wasn't thinking of jwt tbh. but I see what you thought I thought haha. I meant userEnv and Request are two separate entities, and we set the "final env" by merging these two. I happened to call the "final env" -for the lack of a better word- claim.
another thing that I'm struggling with is the existence of userEnv, I can see some key, values as userEnv in the examples and tests but I cannot think of any real life use cases. so to keep things simple and straightforward I suggest either get rid of userEnv altogether or pick a better word for the "final env", if claim is misleading.

Reply:
Aha. I'd say get rid of the userEnv argument (in the Leges.Match function). Because you have already added Leges.definitions, and that's sufficient for my usecases as well.

Reply:
I meant userEnv/definitions in general, I've already removed it in the Leges.Match method